### PR TITLE
build: nuke everything on failure

### PR DESCRIPTION
**Pull request by dt**

### CODE REVIEW

**Comment by benesch** [comment]
IIRC I don't think TeamCity agents can passwordlessly `sudo`. 😞

**Comment by benesch** [comment]
LGTM, modulo running a manual verify-archive build against this branch and verifying the agent is unbricked!

**Comment by benesch** [comment]
https://teamcity.cockroachdb.com/viewQueued.html?itemId=326906&tab=queuedBuildOverviewTab

File: build/teamcity-assert-clean.sh [hunk 1]
```diff
@@ -3,4 +3,7 @@ set -exuo pipefail
 
 export BUILDER_HIDE_GOPATH_SRC=1
 
-build/builder.sh /bin/bash -c '! git status --porcelain | read || (git status; git diff -a 1>&2; exit 1)'
+if ! build/builder.sh /bin/bash -c '! git status --porcelain | read || (git status; git diff -a 1>&2; exit 1)';
```
Comment (benesch) [inline build/teamcity-assert-clean.sh]: you missed a `then`

**Comment by dt** [comment]
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[build/teamcity-assert-clean.sh, line 6 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17720#-KrkiX8t5O5vXTILftO6-r1-6:-KrknMSk9MhbQ2lExSDO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac6df9ce8d7c9cbae152905af461fa2d568322f0/build/teamcity-assert-clean.sh#L6)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

you missed a `then`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17720)*

**Comment by tamird** [comment]
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[build/teamcity-assert-clean.sh, line 6 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17720#-KrkiX8t5O5vXTILftO6-r1-6:-KrktWDwSASN-z3W-nqz:bvtpgvw) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac6df9ce8d7c9cbae152905af461fa2d568322f0/build/teamcity-assert-clean.sh#L6)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Can't this whole thing be:
```
build/builder.sh /bin/bash -c 'git status --porcelain | read || (git status; git diff -a 1>&2; git clean -dxf; exit 1)
```
?
But also, how can the builder remove files owned by root?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17720)*

**Comment by dt** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[build/teamcity-assert-clean.sh, line 6 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17720#-KrkiX8t5O5vXTILftO6-r1-6:-Krkwe7k5mBdKkXLjwQ4:b-7jiiby) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac6df9ce8d7c9cbae152905af461fa2d568322f0/build/teamcity-assert-clean.sh#L6)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Can't this whole thing be:
```
build/builder.sh /bin/bash -c 'git status --porcelain | read || (git status; git diff -a 1>&2; git clean -dxf; exit 1)
```
?
But also, how can the builder remove files owned by root?
</blockquote></details>

it can't: https://teamcity.cockroachdb.com/viewLog.html?buildId=326977&tab=buildResultsDiv&buildTypeId=Cockroach_UnitTests_VerifyArchive

we still need to stop writing as root, but it can nuke whatever it is able to.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17720)*

**Comment by benesch** [comment]
It occurs to me (and @jordanlewis) that the builder doesn't have `sudo`.

**Review by benesch — APPROVED** [review]
:tada: lgtm

**Comment by dt** [comment]
It works! https://teamcity.cockroachdb.com/viewLog.html?buildId=327137&buildTypeId=Cockroach_UnitTests_VerifyArchive&tab=buildLog

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17720#-:-KrlCzC0cc6Egt1FMSWI:b826nkw)*

**Comment by dt** [comment]
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[build/teamcity-assert-clean.sh, line 6 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17720#-KrkiX8t5O5vXTILftO6-r1-6:-KrlOKYixOUceDvNhDgm:bqog9tv) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac6df9ce8d7c9cbae152905af461fa2d568322f0/build/teamcity-assert-clean.sh#L6)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

it can't: https://teamcity.cockroachdb.com/viewLog.html?buildId=326977&tab=buildResultsDiv&buildTypeId=Cockroach_UnitTests_VerifyArchive

we still need to stop writing as root, but it can nuke whatever it is able to.
</blockquote></details>

switched `docker run` which _does_ run it as root

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17720)*

**Pull request merged**